### ISSUE TITLE
linking: Proposal for new visibility attribute "exported"

### ISSUE THREAD

**Issue by sbc100** [WebAssembly/tool-conventions#64]
Right now we support two different visibility options: "hidden" and "default".

"hidden" symbols are not exported unless explicitly exported with `--export`.
"default" symbols are exported either via `-export-dynamic` or individually via `--export` (Or by default if building a DLL.  We currently don't support DLLs but that is the idea).

In emscripten we need a way to implement the EMSCRIPTEN_KEEP_ALIVE macros that is more precise than the current `__attribute__ ((visibility ("default")))`.  This currently means we are passing `-export-dynamic`, which has the side effect of exports all default visibility symbols (i.e. most of libc and libcxx) which means these symbols don't get GCd.

Basically we want stronger idea exporting, something like `__attribute__ ((visibility ("exported")))` which tell the linker to always export a certain symbol.

**Comment by sbc100** [comment]
Or we could just make a separate attribute such as "__attribute__(("wasm_export"))".

**Comment by dschuff** [comment]
I think this makes sense to have. I think this is a third kind of visibility rather than a separate attribute (e.g. it doesn't make sense to have both a visibility of default/hidden and wasm_export, does it?). 
I guess if we were to make a matrix of "is a symbol exported from the wasm binary by default" it would be something like this:

| Visibility        | Static or dynamic binary   | dylib  |
| ------------- |:-------------:| -----:|
| hidden      | No | No |
| default      |  No  |  Yes |
| exported | Yes   |   Yes |

And to summarize my understanding of C attributes and flags:
* The clang flag `-fvisibility` currently sets the default (per object file) to either default or hidden (should we also allow it to be "exported" since it's just per object?)
* The C/C++ attribute syntax overrides the `-fvisibility` setting per-symbol.
* The `--export=foo` linker flag forces `foo` to have "exported" visibility, overriding the object file's definition.
* The `--export-dynamic` linker flag forces all symbols in the linked binary to be exported.

Did I miss anything?

**Comment by sbc100** [comment]
-export-dynamic doesn't apply to hidden symbols.  It basically says "export symbols as if you were building a DSO rather than an executable" which means export anything with default visibility.

**Comment by sbc100** [comment]
I was looking at how to implement this in clang and it seems that adding a new visibility that is only available under wasm might be hard.   This stuff is all defined in `include/clang/Basic/Attr.td`.  There are arch specific attributes but I'm not sure we can have an arch specific setting for the "visibility" attribute.

**Comment by dschuff** [comment]
This would probably object-format-specific rather than arch-specific. Are there arch- or object-format-specific validation rules anywhere (e.g. what happens if you set a visibility on Windows/COFF?)

**Comment by sunfishcode** [comment]
I'm nervous about allowing "hidden" symbols to be exported via linker flags. The optimizer considers itself permitted to change the signature and/or calling convention of "hidden" symbols, such as deleting unneeded arguments, because it believes it can see all the code which could reference them by name.

Concerning module-exported attributes, my initial sense here is that it makes sense to introduce a wasm-specific attribute here, in addition to the existing visibility attribute. I'm not aware of another level of symbol scoping above dynamic libraries on other platforms.

On the topic of dynamic linking, my understanding is that dynamic libraries will want to *be* wasm modules, because wasm modules are the unit of code that a wasm engine can easily load. This would mean no inherent distinction between "visible outside the dynamic library" and "visible outside the module". If we want to support exporting only a restricted view of a module to JS etc., a concept which may be useful is a separate interface module, which would import from the C++ dynamic library (or libraries) and then only export those things which make sense outside.

**Comment by sbc100** [comment]
@sunfishcode Regarding your first point, my understanding is that visibility is used by the linker, not the compiler.  Perhaps you are thinking of the symbol binding ("local" vs "global")?   Put another way:  visibility hidden symbols can still be referenced by other translation units, just not of DSOs.

**Comment by sunfishcode** [comment]
@sbc100 Oh, you're right. LTO can internalize hidden symbols and then do those optimizations, but from that point on that point on they're local. That said, I just did a quick grep and found [this](https://github.com/llvm-mirror/llvm/blob/master/lib/Analysis/InstructionSimplify.cpp#L2303), which I'm not sure what to say about.

**Comment by NWilson** [comment]
People do manage with only hidden/default visibilities, on other platforms. It relies on knowing at compile-time whether or not you're creating a binary or a DSO: for DSOs, you compile your interface with `-fvisibility=default` so it gets exported; for a binary you just compile all the code you link in using `-fvisibility=hidden` to avoid clogging up your binary's symbol table (if you even care). So existing toolchains for building applications and libraries are expecting the current system.

Who are the main target audience for the new proposed attribute/visibility-level? Is it mainly for library authors, who want to make sure that their library's interface is exported when compiled as part of a larger static Wasm module, but without forcing the person linking the module to export all default-visibility symbols? How common a use-case is it? There is an existing workaround for this situation: libraries provide headers that apply default-visibility to their symbols, and people building modules make sure to apply `-fvisibility=hidden` when compiling the code that they plan to statically link in to their module.

**Comment by sbc100** [comment]
Yes you are correct about the main use case.  Specifically emscripten has EMSCRIPTEN_KEEP_ALIVE which means that symbol is needed by JS.   Currently we use -export-dynamic to export all the default visibility symbols .. but that result is all kind of standard library symbols (libc/libc++) being exported too which we don't want.

An alternative, as you say is to audit all the libraries and make sure they are not marking any symbols as "default", but that is error prone and imprecise, and means we would need a second recompiled version of the the libraries when we are building for dynamic linking since in the dynamic linking case that libraries should be setting the default visibility as needed.   We could take this approach though I suppose.

You can think of default visibility as "make this symbol visible to other DSOs" and this new thing as "make this symbol available to the embedder" I guess.

**Comment by dschuff** [comment]
@NWilson the use case here is for controlling exports from a linked binary (be it statically or dynamically linked). These exports have to exist in order to be called from JS code, but they have a a binary size and instantiation-time cost so we don't want to export everything by default. On the web even (most) statically-linked binaries are in some sense dynamic libraries. On other platforms there isn't much use in exporting symbols from statically-linked binaries (other than a few obscure use cases, for which `--export-dynamic` suffices).

**Comment by NWilson** [comment]
I see, so it's sort of an override to aid in managing Emscripten's ports repository, to avoid having to worry about whether the libraries are hiding/exporting their symbols as expecting. While that would be the normal management approach, it's too much effort to police it in the Emscripten ports on a library-by-library basis, so the new flag allows exporting some symbols without need to opt in to exporting all default-visibility symbols.

You just have to hope you don't start and arms race! If library vendors get the message that "wasm_export" is the Wasm-y way to export things, they might start using instead of default-visibility on Wasm :(

**Comment by sbc100** [comment]
Thats a good point.  Overuse of that attribute would be make us sad.

We could investigate trying configure emscriptens libraries for static linking first and see how far that gets us before we decide we really need this feature.

**Comment by dschuff** [comment]
I see it as less about managing emscripten's library ports and more about allowing the user of the libraries control the exports of the final binary+library product. Library authors should probably just behave the same as they would on other platforms. I don't think they should be using EMSCRIPTEN_KEEPALIVE. They would typically use dllexport or default visibility to mark their own exports (to their library users), and the library user who links that into a binary would then decide what to export from the final binary. I guess the interesting case is someone who really did just want to build only a C/C++ library and expose it directly to JS with no C code in between, i.e. overlapping the use cases. I sitll feel like that should be a distinct link command (different from what you would use to link a dylib) and it can have different behavior.

**Comment by sbc100** [comment]
I think that overlapping use case you describe is basically '-Wl,-export-dynamic' which basically means "export from my executable anything that is visibility default".

**Comment by kripken** [comment]
A related issue may be lld's handling of `llvm.used` (which is how emscripten implements `EMSCRIPTEN_KEEPALIVE`). The LLVM langref says

> If a symbol appears in the @llvm.used list, then the compiler, assembler, and linker are required to treat the symbol as if there is a reference to the symbol that it cannot see (which is why they have to be named). For example, if a variable has internal linkage and no references other than that from the @llvm.used list, it cannot be deleted.

It looks like lld behaves as follows:

 * no flags: the function is not in the output
 * `--no-gc-sections`: the function is in the output.
 * `--export-dynamic', with or without `--no-gc-sections`: the function is in the output and is exported.

@sbc100 is that expected?

If `lld` instead exported the functions in `llvm.used`, that would both preserve them through `gc-sections` and also might be enough to build `EMSCRIPTEN_KEEPALIVE` on top of. But perhaps that's not the intention?

**Comment by sbc100** [comment]
Yes that is exactly the behaviour I would expect and IIUC that matches what the native GNU linker would do.

I like your idea about using the existing 'llvm.used' attribute here.  I think I will plumb that through:https://reviews.llvm.org/D57864

**Comment by sbc100** [comment]
Updated llvm to honor llvm.used.

lld change is in flight: https://reviews.llvm.org/D57869

Once that lands we can remove `--export-dynamic` from the emscripten

**Comment by corwin-of-amber** [comment]
How was this resolved, then, eventually? `llvm.used` is an internal structure; is there any way to affect it using an `__attribute__` like in the original proposal?

**Comment by corwin-of-amber** [comment]
Thanks. I just tried it with [wasi-sdk](https://github.com/CraneStation/wasi-sdk) and it seems to work there as well.

**Comment by sbc100** [comment]
The `used` attribute should not translate into an export with the wasi-sdk.  At least not with the latest llvm.  just FYI.

**Comment by corwin-of-amber** [comment]
Was it introduced unintentionally into wasi-sdk then? Because I have compiled the following program:
```c
__attribute__((used)) void issue64() { }

int main(int argc, char *argv[]) {
    return 0;
}
```

Using wasi-sdk's Clang:
```sh
/opt/wasi-sdk/bin/clang sbc100.c
wasm2wat --dir=. a.out | grep issue
```

And I clearly see
```
  (func $issue64 (type 3)
  (export "issue64" (func $issue64))
```

**Comment by sunfishcode** [comment]
The patch to fix this landed here: https://reviews.llvm.org/D62542, which is more recent than LLVM 9.0, so it's not in the current wasi-sdk.

**Comment by cnlohr** [comment]
This does not appear to be working I'm on clang 10.1, and even if I have

```c
int main() __attribute__((used))
```
```bash
clang -Irawdraw -DWASM -nostdlib --target=wasm32 -flto -Wl,--lto-O3 -Wl,--no-entry -Wl,--allow-undefined -Wl,--import-memory -Wl,--strip-all rawdraw.c -o main.wasm
```

It does not persist to the output.  I do not want to have to specify every symbol I want to export on the command-line.

**Comment by cnlohr** [comment]
Ah! Found it

```c
int __attribute__((export_name("main"))) main()
```
seems to work wonders! 

```
clang -Irawdraw -DWASM -nostdlib --target=wasm32 -flto -Wl,--lto-O3 -Wl,--no-entry -Wl,--allow-undefined -Wl,--import-memory rawdraw.c -o main.wasm
```

**Comment by fwsGonzo** [comment]
Just pass --undefined=symbolname to the linker. Works in every language, not just WASM. You don't have to export every symbol in the whole glibc just to export a single function.

**Comment by sbc100** [comment]
I don't think `--undefined=symbolname` actually exports anything.   It simply forces the symbol to be pulled out of an archive file if if it part of one.   IIRC it doesn't even cause the link to fail if the symbol is missing.. for that you would need `--require-defined=symbolname`.

Under wasm neither `--undefined`, nor `--require-defined` cause anything to be exported.  In order to actually export the symbol you would need to use either `--export` or `--export-if-defined` (which are both wasm-specific linker flags).

**Comment by fwsGonzo** [comment]
The problem for me has been to find something that works for static executables that have LTO and section GC. Using --undefined= is the only thing that works so far for me in any language, on every linker, and not just WebAssembly.

If you have added something new that is specific to wasm, that is great. I'll make a mental note. :+1:
